openvpn.8
.\"  OpenVPN -- An application to securely tunnel IP networks
.\"             over a single TCP/UDP port, with support for SSL/TLS-based
.\"             session authentication and key exchange,
.\"             packet encryption, packet authentication, and
.\"             packet compression.
.\"
.\"  Copyright (C) 2002-2010 OpenVPN Technologies, Inc. <sales@openvpn.net>
.\"
.\"  This program is free software; you can redistribute it and/or modify
.\"  it under the terms of the GNU General Public License version 2
.\"  as published by the Free Software Foundation.
.\"
.\"  This program is distributed in the hope that it will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"
.\"  You should have received a copy of the GNU General Public License
.\"  along with this program (see the file COPYING included with this
.\"  distribution); if not, write to the Free Software Foundation, Inc.,
.\"  59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.\"
.\" Manual page for openvpn
.\
.\" SH section heading
.\" SS subsection heading
.\" LP paragraph
.\" IP indented paragraph
.\" TP hanging label
.\
.\" .nf -- no formatting
.\" .fi -- resume formatting
.\" .ft 3 -- boldface
.\" .ft -- normal face
.\" .in +|-{n} -- indent
.\"
.TH openvpn 8 "17 November 2008"
.\"*********************************************************
.SH NAME
openvpn \- secure IP tunnel daemon.
.\"*********************************************************
.SH SYNOPSIS
.ft 3
openvpn [ options ... ]
.ft
.\"*********************************************************
.SH INTRODUCTION
.LP
OpenVPN is an open source VPN daemon by James Yonan.
Because OpenVPN tries to
be a universal VPN tool offering a great deal of flexibility,
there are a lot of options on this manual page.
If you're new to OpenVPN, you might want to skip ahead to the
examples section where you will see how to construct simple
VPNs on the command line without even needing a configuration file.

Also note that there's more documentation and examples on
the OpenVPN web site:
.I http://openvpn.net/

And if you would like to see a shorter version of this manual,
see the openvpn usage message which can be obtained by
running
.B openvpn
without any parameters.
.\"*********************************************************
.SH DESCRIPTION
.LP
OpenVPN is a robust and highly flexible VPN daemon.
OpenVPN supports SSL/TLS security, ethernet bridging,
TCP or UDP tunnel transport through proxies or NAT,
support for dynamic IP addresses and DHCP,
scalability to hundreds or thousands of users,
and portability to most major OS platforms.

OpenVPN is tightly bound to the OpenSSL library, and derives much
of its crypto capabilities from it.

OpenVPN supports
conventional encryption
using a pre-shared secret key
.B (Static Key mode)
or
public key security
.B (SSL/TLS mode)
using client & server certificates.
OpenVPN also
supports non-encrypted TCP/UDP tunnels.  

OpenVPN is designed to work with the
.B TUN/TAP
virtual networking interface that exists on most platforms.

Overall, OpenVPN aims to offer many of the key features of IPSec but
with a relatively lightweight footprint.
.\"*********************************************************
.SH OPTIONS
OpenVPN allows any option to be placed either on the command line
or in a configuration file.  Though all command line options are preceded
by a double-leading-dash ("\-\-"), this prefix can be removed when
an option is placed in a configuration file.
.\"*********************************************************
.TP
.B \-\-help
Show options.
.\"*********************************************************
.TP
.B \-\-config file
Load additional config options from
.B file
where each line corresponds to one command line option,
but with the leading '\-\-' removed.

If
.B \-\-config file
is the only option to the openvpn command,
the
.B \-\-config
can be removed, and the command can be given as
.B openvpn file

Note that
configuration files can be nested to a reasonable depth.

Double quotation or single quotation characters ("", '')
can be used to enclose single parameters containing whitespace,
and "#" or ";" characters in the first column
can be used to denote comments.

Note that OpenVPN 2.0 and higher performs backslash-based shell
escaping for characters not in single quotations,
so the following mappings should be observed:

.nf
.ft 3
.in +4
\\\\       Maps to a single backslash character (\\).
\\"       Pass a literal doublequote character ("), don't
         interpret it as enclosing a parameter.
\\[SPACE] Pass a literal space or tab character, don't
         interpret it as a parameter delimiter.
.in -4
.ft
.fi

For example on Windows, use double backslashes to
represent pathnames:

.nf
.ft 3
.in +4
secret "c:\\\\OpenVPN\\\\secret.key"
.in -4
.ft
.fi

For examples of configuration files,
see
.I http://openvpn.net/examples.html

Here is an example configuration file:

.nf
.ft 3
.in +4
#
# Sample OpenVPN configuration file for
# using a pre-shared static key.
#
# '#' or ';' may be used to delimit comments.

# Use a dynamic tun device.
dev tun

# Our remote peer
remote mypeer.mydomain

# 10.1.0.1 is our local VPN endpoint
# 10.1.0.2 is our remote VPN endpoint
ifconfig 10.1.0.1 10.1.0.2

# Our pre-shared static key
secret static.key
.in -4
.ft
.fi
.\"*********************************************************
.SS Tunnel Options:
.TP
.B \-\-mode m
Set OpenVPN major mode.  By default, OpenVPN runs in
point-to-point mode ("p2p").  OpenVPN 2.0 introduces
a new mode ("server") which implements a multi-client
server capability.
.\"*********************************************************
.TP
.B \-\-local host
Local host name or IP address for bind.
If specified, OpenVPN will bind to this address only.
If unspecified, OpenVPN will bind to all interfaces.
.\"*********************************************************
.TP
.B \-\-remote host [port] [proto]
Remote host name or IP address.  On the client, multiple
.B \-\-remote
options may be specified for redundancy, each referring
to a different OpenVPN server.  Specifying multiple
.B \-\-remote
options for this purpose is a special case of the more
general connection-profile feature.  See the
.B <connection>
documentation below.

The OpenVPN client will try to connect to a server at
.B host:port
in the order specified by the list of
.B \-\-remote
options.

.B proto
indicates the protocol to use when connecting with the
remote, and may be "tcp" or "udp".

The client will move on to the next host in the list,
in the event of connection failure.
Note that at any given time, the OpenVPN client
will at most be connected to
one server.

Note that since UDP is connectionless, connection failure
is defined by the
.B \-\-ping
and
.B \-\-ping-restart
options.

Note the following corner case:  If you use multiple
.B \-\-remote
options, AND you are dropping root privileges on
the client with
.B \-\-user
and/or
.B \-\-group,
AND the client is running a non-Windows OS, if the client needs
to switch to a different server, and that server pushes
back different TUN/TAP or route settings, the client may lack
the necessary privileges to close and reopen the TUN/TAP interface.
This could cause the client to exit with a fatal error.

If
.B \-\-remote
is unspecified, OpenVPN will listen
for packets from any IP address, but will not act on those packets unless
they pass all authentication tests.  This requirement for authentication
is binding on all potential peers, even those from known and supposedly
trusted IP addresses (it is very easy to forge a source IP address on
a UDP packet).

When used in TCP mode, 
.B \-\-remote
will act as a filter, rejecting connections from any host which does
not match
.B host.

If
.B host
is a DNS name which resolves to multiple IP addresses,
one will be randomly
chosen, providing a sort of basic load-balancing and
failover capability.
.\"*********************************************************
.TP
.B \-\-remote-random-hostname
Add a random string (6 characters) to first DNS label of hostname to prevent
DNS caching.  For example, "foo.bar.gov" would be modified to
"<random-chars>.foo.bar.gov".
.\"*********************************************************
.TP
.B <connection>
Define a client connection
profile.  Client connection profiles are groups of OpenVPN options that
describe how to connect to a given OpenVPN server.  Client connection
profiles are specified within an OpenVPN configuration file, and
each profile is bracketed by
.B <connection>
and
.B </connection>.

An OpenVPN client will try each connection profile sequentially
until it achieves a successful connection.  

.B \-\-remote-random
can be used to initially "scramble" the connection
list.

Here is an example of connection profile usage:

.nf
.ft 3
.in +4
client
dev tun

<connection>
remote 198.19.34.56 1194 udp
</connection>

<connection>
remote 198.19.34.56 443 tcp
</connection>

<connection>
remote 198.19.34.56 443 tcp
http-proxy 192.168.0.8 8080
http-proxy-retry
</connection>

<connection>
remote 198.19.36.99 443 tcp
http-proxy 192.168.0.8 8080
http-proxy-retry
</connection>

persist-key
persist-tun
pkcs12 client.p12
ns-cert-type server
verb 3
.in -4
.ft
.fi

First we try to connect to a server at 198.19.34.56:1194 using UDP.
If that fails, we then try to connect to 198.19.34.56:443 using TCP.
If that also fails, then try connecting through an HTTP proxy at 
192.168.0.8:8080 to 198.19.34.56:443 using TCP.  Finally, try to
connect through the same proxy to a server at 198.19.36.99:443
using TCP.

The following OpenVPN options may be used inside of
a
.B <connection>
block:

.B bind,
.B connect-retry,
.B connect-retry-max,
.B connect-timeout,
.B float,
.B http-proxy,
.B http-proxy-option,
.B http-proxy-retry,
.B http-proxy-timeout,
.B local,
.B lport,
.B nobind,
.B port,
.B proto,
.B remote,
.B rport,
.B socks-proxy, and
.B socks-proxy-retry.

A defaulting mechanism exists for specifying options to apply to
all
.B <connection>
profiles.  If any of the above options (with the exception of
.B remote
) appear outside of a
.B <connection>
block, but in a configuration file which has one or more
.B <connection>
blocks, the option setting will be used as a default for
.B <connection>
blocks which follow it in the configuration file.

For example, suppose the
.B nobind
option were placed in the sample configuration file above, near
the top of the file, before the first
.B <connection>
block.  The effect would be as if
.B nobind
were declared in all
.B <connection>
blocks below it.
.\"*********************************************************
.TP
.B \-\-proto-force p
When iterating through connection profiles,
only consider profiles using protocol
.B p
('tcp'|'udp'). 
.\"*********************************************************
.TP
.B \-\-remote-random
When multiple
.B \-\-remote
address/ports are specified, or if connection profiles are being
used, initially randomize the order of the list
as a kind of basic load-balancing measure.
.\"*********************************************************
.TP
.B \-\-proto p
Use protocol
.B p
for communicating with remote host.
.B p
can be
.B udp,
.B tcp-client,
or
.B tcp-server.

The default protocol is
.B udp
when
.B \-\-proto
is not specified.

For UDP operation,
.B \-\-proto udp
should be specified on both peers.

For TCP operation, one peer must use
.B \-\-proto tcp-server
and the other must use
.B \-\-proto tcp-client.
A peer started with
.B tcp-server
will wait indefinitely for an incoming connection.  A peer
started with
.B tcp-client
will attempt to connect, and if that fails, will sleep for 5
seconds (adjustable via the
.B \-\-connect-retry
option) and try again infinite or up to N retries (adjustable via the
.B \-\-connect-retry-max
option).  Both TCP client and server will simulate
a SIGUSR1 restart signal if either side resets the connection.

OpenVPN is designed to operate optimally over UDP, but TCP capability is provided
for situations where UDP cannot be used.
In comparison with UDP, TCP will usually be
somewhat less efficient and less robust when used over unreliable or congested
networks.

This article outlines some of problems with tunneling IP over TCP:

.I http://sites.inka.de/sites/bigred/devel/tcp-tcp.html

There are certain cases, however, where using TCP may be advantageous from
a security and robustness perspective, such as tunneling non-IP or
application-level UDP protocols, or tunneling protocols which don't
possess a built-in reliability layer.
.\"*********************************************************
.TP
.B \-\-connect-retry n
For
.B \-\-proto tcp-client,
take
.B n
as the
number of seconds to wait
between connection retries (default=5).
.\"*********************************************************
.TP
.B \-\-connect-timeout n
For
.B \-\-proto tcp-client,
set connection timeout to
.B n
seconds (default=10).
.\"*********************************************************
.TP
.B \-\-connect-retry-max n
For
.B \-\-proto tcp-client,
take
.B n
as the
number of retries of connection attempt (default=infinite).
.\"*********************************************************
.TP
.B \-\-auto-proxy
Try to sense HTTP or SOCKS proxy settings automatically.
If no settings are present, a direct connection will be attempted.
If both HTTP and SOCKS settings are present, HTTP will be preferred.
If the HTTP proxy server requires a password, it will be queried from
stdin or the management interface.  If the underlying OS doesn't support an API for
returning proxy settings, a direct connection will be attempted.
Currently, only Windows clients support this option via the
InternetQueryOption API.
This option exists in OpenVPN 2.1 or higher.
.\"*********************************************************
.TP
.B \-\-show-proxy-settings
Show sensed HTTP or SOCKS proxy settings. Currently, only Windows clients
support this option.
.\"*********************************************************
.TP
.B \-\-http-proxy server port [authfile|'auto'|'auto-nct'] [auth-method]
Connect to remote host through an HTTP proxy at address
.B server
and port
.B port.
If HTTP Proxy-Authenticate is required,
.B authfile
is a file containing a username and password on 2 lines, or
"stdin" to prompt from console.

.B auth-method
should be one of "none", "basic", or "ntlm".

HTTP Digest authentication is supported as well, but only via
the
.B auto
or
.B auto-nct
flags (below).

The
.B auto
flag causes OpenVPN to automatically determine the
.B auth-method
and query stdin or the management interface for
username/password credentials, if required.  This flag
exists on OpenVPN 2.1 or higher.

The
.B auto-nct
flag (no clear-text auth) instructs OpenVPN to automatically
determine the authentication method, but to reject weak
authentication protocols such as HTTP Basic Authentication.
.\"*********************************************************
.TP
.B \-\-http-proxy-retry
Retry indefinitely on HTTP proxy errors.  If an HTTP proxy error
occurs, simulate a SIGUSR1 reset.
.\"*********************************************************
.TP
.B \-\-http-proxy-timeout n
Set proxy timeout to
.B n
seconds, default=5.
.\"*********************************************************
.TP
.B \-\-http-proxy-option type [parm]
Set extended HTTP proxy options.
Repeat to set multiple options.

.B VERSION version \-\-
Set HTTP version number to
.B version
(default=1.0).

.B AGENT user-agent \-\-
Set HTTP "User-Agent" string to
.B user-agent.
.\"*********************************************************
.TP
.B \-\-socks-proxy server [port]
Connect to remote host through a Socks5 proxy at address
.B server
and port
.B port
(default=1080).
.\"*********************************************************
.TP
.B \-\-socks-proxy-retry
Retry indefinitely on Socks proxy errors.  If a Socks proxy error
occurs, simulate a SIGUSR1 reset.
.\"*********************************************************
.TP
.B \-\-resolv-retry n
If hostname resolve fails for
.B \-\-remote,
retry resolve for
.B n
seconds before failing.

Set
.B n
to "infinite" to retry indefinitely.

By default,
.B \-\-resolv-retry infinite
is enabled.  You can disable by setting n=0.
.\"*********************************************************
.TP
.B \-\-float
Allow remote peer to change its IP address and/or port number, such as due to
DHCP (this is the default if
.B \-\-remote
is not used).
.B \-\-float
when specified with
.B \-\-remote
allows an OpenVPN session to initially connect to a peer
at a known address, however if packets arrive from a new
address and pass all authentication tests, the new address
will take control of the session.  This is useful when
you are connecting to a peer which holds a dynamic address
such as a dial-in user or DHCP client.

Essentially,
.B \-\-float
tells OpenVPN to accept authenticated packets
from any address, not only the address which was specified in the
.B \-\-remote
option.
.\"*********************************************************
.TP
.B \-\-ipchange cmd
Execute shell command
.B cmd
when our remote ip-address is initially authenticated or
changes.

Execute as:

.B cmd ip_address port_number

Don't use
.B \-\-ipchange
in
.B \-\-mode server
mode.  Use a
.B \-\-client-connect
script instead.

See the "Environmental Variables" section below for
additional parameters passed as environmental variables.

Note that
.B cmd
can be a shell command with multiple arguments, in which
case all OpenVPN-generated arguments will be appended
to
.B cmd
to build a command line which will be passed to the script.

If you are running in a dynamic IP address environment where
the IP addresses of either peer could change without notice,
you can use this script, for example, to edit the
.I /etc/hosts
file with the current address of the peer.  The script will
be run every time the remote peer changes its IP address.

Similarly if
.I our
IP address changes due to DHCP, we should configure
our IP address change script (see man page for
.BR dhcpcd (8)
) to deliver a
.B SIGHUP
or
.B SIGUSR1
signal to OpenVPN.  OpenVPN will then
reestablish a connection with its most recently authenticated
peer on its new IP address.
.\"*********************************************************
.TP
.B \-\-port port
TCP/UDP port number for both local and remote.  The current
default of 1194 represents the official IANA port number
assignment for OpenVPN and has been used since version 2.0-beta17.
Previous versions used port 5000 as the default.
.\"*********************************************************
.TP
.B \-\-lport port
TCP/UDP port number for bind.
.\"*********************************************************
.TP
.B \-\-rport port
TCP/UDP port number for remote.
.\"*********************************************************
.TP
.B \-\-bind
Bind to local address and port. This is the default unless any of 
.B \-\-proto tcp-client
,
.B \-\-http-proxy
or
.B \-\-socks-proxy
are used.
.\"*********************************************************
.TP
.B \-\-nobind
Do not bind to local address and port.  The IP stack will allocate
a dynamic port for returning packets.  Since the value of the dynamic port
could not be known in advance by a peer, this option is only suitable for
peers which will be initiating connections by using the
.B \-\-remote
option.
.\"*********************************************************
.TP
.B \-\-dev tunX | tapX | null
TUN/TAP virtual network device (
.B X
can be omitted for a dynamic device.)

See examples section below
for an example on setting up a TUN device.

You must use either tun devices on both ends of the connection
or tap devices on both ends.  You cannot mix them, as they
represent different underlying network layers.

.B tun
devices encapsulate IPv4 or IPv6 (OSI Layer 3) while
.B tap
devices encapsulate Ethernet 802.3 (OSI Layer 2).
.\"*********************************************************
.TP
.B \-\-dev-type device-type
Which device type are we using?
.B device-type
should be
.B tun
(OSI Layer 3)
or
.B tap
(OSI Layer 2).
Use this option only if the TUN/TAP device used with
.B \-\-dev
does not begin with
.B tun
or
.B tap.
.\"*********************************************************
.TP
.B \-\-topology mode
Configure virtual addressing topology when running in
.B \-\-dev tun
mode.  This directive has no meaning in
.B \-\-dev tap
mode, which always uses a
.B subnet
topology.

If you set this directive on the server, the
.B \-\-server
and
.B \-\-server-bridge
directives will automatically push your chosen topology setting to clients
as well.  This directive can also be manually pushed to clients.  Like the
.B \-\-dev
directive, this directive must always be compatible between client and server.

.B mode
can be one of:

.B net30 \-\-
Use a point-to-point topology, by allocating one /30 subnet per client.
This is designed to allow point-to-point semantics when some
or all of the connecting clients might be Windows systems.  This is the
default on OpenVPN 2.0.

.B p2p \-\-
Use a point-to-point topology where the remote endpoint of the client's
tun interface always points to the local endpoint of the server's tun interface.
This mode allocates a single IP address per connecting client.
Only use
when none of the connecting clients are Windows systems.  This mode

ChangeLog
OpenVPN
Copyright (C) 2002-2005 OpenVPN Solutions LLC <info@openvpn.net>

$Id: ChangeLog,v 1.128 2005/04/11 04:16:14 jimyonan Exp $

2005.04.17 -- Version 2.0

* Fixed minor options string typo in options.c.

2005.04.10 -- Version 2.0-rc21

* Change license description from "GPL Version 2 or (at your
  option) any later version" to just "GPL Version 2".

2005.04.04 -- Version 2.0-rc20

* Dag Wieers has put together an OpenVPN/LZO binary RPM set with
  excellent distro/version coverage for RH/EL/Fedora, though
  using his own SPEC.  I modified openvpn.spec to follow some of
  the same conventions such as putting sample scripts and doc
  files in %doc rather than /usr/share/openvpn.
* Minor change to init scripts to run the user-defined script
  /etc/openvpn/openvpn-startup (if it exists) before any OpenVPN
  configs are started, and to run /etc/openvpn/openvpn-shutdown
  after all OpenVPN configs have been stopped.  The
  openvpn-startup script can be used for stuff like
  insmod tun.o, setting up firewall rules, or starting
  ethernet bridges.

2005.03.29 -- Version 2.0-rc19

* Omit additions of routes where the network and
  gateway are equal and the netmask is 255.255.255.255.
  This can come up if you are using both
  server/ifconfig-pool and client-config-dir with
  ifconfig-push static addresses for some subset of clients
  which directly reference the server IP address as the
  remote endpoint.
	
2005.03.28 -- Version 2.0-rc18

* Packaged Windows installer with OpenSSL 0.9.7f.
* Built Windows installer with NSIS 2.06.

2005.03.12 -- Version 2.0-rc17

* "MANAGEMENT: CMD" log file output will now only occur
  at --verb 7 or greater.
* Added an optional name/value configuration list to
  the openvpn-auth-pam plugin module argument list.  See
  plugin/auth-pam/README for documentation. This is necessary
  in order for openvpn-auth-pam to work with queries generated
  by arbitrary PAM modules.
* In both auth-pam and down-root plugins, in the forked process,
  a read error on the parent process socket is no longer fatal.
* MandrakeSoft liblzo1 RPM only Provides for a 'liblzo1'.
  A conditional test of the vendor has been added to
  Require the appropriately named 'lzo' (liblzo1 / lzo).
  (Tom Walsh - http://openhardware.net)

	
2005.02.20 -- Version 2.0-rc16

* Fixed bug introduced in rc13 where Windows service wrapper
  would be installed with a startup type of Automatic.
  This fix restores the previous behavior of installing
  with a startup type of Manual.

2005.02.19 -- Version 2.0-rc15

* Added warning when --keepalive is not used in a server
  configuration.
* Don't include OpenSSL md4.h file if we are not building
  NTLM proxy support (Waldemar Brodkorb).
* Added easy-rsa/build-key-pkcs12 and
  easy-rsa/Windows/build-key-pkcs12.bat scripts
  (Mathias Sundman).

2005.02.16 -- Version 2.0-rc14

* Fixed small memory leak that occurs when --crl-verify
  is used.
* Upgraded Windows installer and .nsi script to NSIS 2.05
  (Mathias Sundman).
* Changed #include backslash usage in cryptoapi.c to use
  forward slashes instead (Gisle Vanem).
* Created easy-rsa/revoke-full to handle revocations in
  a single step: (a) revoke crt, (b) regenerate CRL, and
  (c) verify that revocation succeeded.
* Renamed easy-rsa/Windows/revoke-key to revoke-full so
  that both *nix and Windows scripts are equivalent.
	
2005.02.11 -- Version 2.0-rc13

* Improve human-readability of local/remote options
  diff, when inconsistencies are present.
* For Windows easy-rsa, distribute vars.bat.sample and
  openssl.cnf.sample, then copy them to their normal
  filenames (without the .sample) when init-config.bat
  is run.  This is to prevent OpenVPN upgrades from
  wiping out vars.bat and openssl.cnf edits.
* Modified service wrapper (Windows) to use a
  case-insensitive search when scanning for .ovpn files
  in \Program Files\OpenVPN\config.  Prior versions
  required an all-lower-case .ovpn file extension.
* Miscellaneous service wrapper code cleanup.
* If --user/--group is used on Windows, treat it
  as a no-op with a warning (this makes it easier to
  distribute the same client config file to Windows
  and *nix users).
* Warn if --ifconfig-pool-persist is used with
  --duplicate-cn.

2005.02.05 -- Version 2.0-rc12

* Removed some debugging code inadvertently included
  in rc11 which would print the --auth-user-pass
  username/password provided by clients in the server
  logfile.
* Client code for cycling through --remote list will
  retry the last address which successfully authenticated
  before moving on through the list.
* Windows installer will now install sample configuration
  files in \Program Files\OpenVPN\sample-configs as well
  as generate a start menu shortcut to this directory.
* Minor type change in buffer.[ch] to work around char-type
  ambiguity bug.  Caused management interface lock-ups on
  ARM when building with armv4b-hardhat-linux-gcc 2.95.3.

2005.02.03 -- Version 2.0-rc11

* Windows installer will now install easy-rsa directory
  in \Program Files\OpenVPN
* Allow syslog facility to be controlled at compile time,
  e.g. -DLOG_OPENVPN=LOG_LOCAL6 (P Kern).
* Changed certain shell scripts in distribution to use
  #!/bin/sh rather than #!/bin/bash for better portability.
* If --ifconfig-pool-persist seconds parameter is 0, treat
  persist file as an allocation of fixed IP addresses
  (previous versions took IP-to-common-name associations
  from this list as hints, not mandatory static allocations).
* Fixed bug on *nix where if --auth-user-pass and --log
  were used together, the username prompt would be sent to
  the log file rather than /dev/tty.
* Spurious text in openvpn.8 detected by doclifter
  (Eric S. Raymond).
* Call closelog later on daemon kill so that process
  exit message is written to syslog.

2005.01.27 -- Version 2.0-rc10

* When ./configure is run with plugins enabled (the default),
  check whether or not dlopen exists in libc before testing

debian/openvpn.init.d
#!/bin/sh -e

### BEGIN INIT INFO
# Provides:          openvpn
# Required-Start:    $network $remote_fs $syslog
# Required-Stop:     $network $remote_fs $syslog
# Should-Start:      network-manager
# Should-Stop:       network-manager
# X-Start-Before:    $x-display-manager gdm kdm xdm wdm ldm sdm nodm
# X-Interactive:     true
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# Short-Description: Openvpn VPN service
### END INIT INFO

# Original version by Robert Leslie
# <rob@mars.org>, edited by iwj and cs
# Modified for openvpn by Alberto Gonzalez Iniesta <agi@inittab.org>
# Modified for restarting / starting / stopping single tunnels by Richard Mueller <mueller@teamix.net>

. /lib/lsb/init-functions

test $DEBIAN_SCRIPT_DEBUG && set -v -x

DAEMON=/usr/sbin/openvpn
DESC="virtual private network daemon(s)"
CONFIG_DIR=/etc/openvpn
test -x $DAEMON || exit 0
test -d $CONFIG_DIR || exit 0

# Source defaults file; edit that file to configure this script.
AUTOSTART="all"
STATUSREFRESH=10
OMIT_SENDSIGS=0
if test -e /etc/default/openvpn ; then
  . /etc/default/openvpn
fi

start_vpn () {
    if grep -q '^[	 ]*daemon' $CONFIG_DIR/$NAME.conf ; then
      # daemon already given in config file
      DAEMONARG=
    else
      # need to daemonize
      DAEMONARG="--daemon ovpn-$NAME"
    fi

    if grep -q '^[	 ]*status ' $CONFIG_DIR/$NAME.conf ; then
      # status file already given in config file
      STATUSARG=""
    elif test $STATUSREFRESH -eq 0 ; then
      # default status file disabled in /etc/default/openvpn
      STATUSARG=""
    else
      # prepare default status file
      STATUSARG="--status /var/run/openvpn.$NAME.status $STATUSREFRESH"
    fi

    # Handle backwards compatibility
    script_security=""
    if test -z "$( grep '^[[:space:]]*script-security[[:space:]]' $CONFIG_DIR/$NAME.conf )" ; then
        script_security="--script-security 2"
    fi

    STATUS=0
    start-stop-daemon --start --quiet --oknodo \
        --pidfile /var/run/openvpn.$NAME.pid \
        --exec $DAEMON -- $OPTARGS --writepid /var/run/openvpn.$NAME.pid \
        $DAEMONARG $STATUSARG --cd $CONFIG_DIR \
        --config $CONFIG_DIR/$NAME.conf < /dev/null || STATUS=1

    [ "$OMIT_SENDSIGS" -ne 1 ] || ln -s /var/run/openvpn.$NAME.pid /run/sendsigs.omit.d/openvpn.$NAME.pid
}
stop_vpn () {
  kill `cat $PIDFILE` || true
  rm -f $PIDFILE
  [ "$OMIT_SENDSIGS" -ne 1 ] || rm -f /run/sendsigs.omit.d/openvpn.$NAME.pid
  rm -f /var/run/openvpn.$NAME.status 2> /dev/null
  log_end_msg 0
}

case "$1" in
start)
  log_action_begin_msg "Starting $DESC"

  # autostart VPNs
  if test -z "$2" ; then
    # check if automatic startup is disabled by AUTOSTART=none
    if test "x$AUTOSTART" = "xnone" -o -z "$AUTOSTART" ; then
      log_warning_msg "  Autostart disabled, no VPN will be started."
      exit 0
    fi
    if test -z "$AUTOSTART" -o "x$AUTOSTART" = "xall" ; then
      # all VPNs shall be started automatically
      for CONFIG in `cd $CONFIG_DIR; ls *.conf 2> /dev/null`; do
        NAME=${CONFIG%%.conf}
        log_daemon_msg "  Autostarting VPN '$NAME'"
        start_vpn
      done
    else
      # start only specified VPNs
      for NAME in $AUTOSTART ; do
        if test -e $CONFIG_DIR/$NAME.conf ; then
          log_daemon_msg "  Autostarting VPN '$NAME'"
          start_vpn
        else
          log_failure_msg "  Autostarting VPN '$NAME': missing $CONFIG_DIR/$NAME.conf file !"
          STATUS=1
        fi
      done
    fi
  #start VPNs from command line
  else
    while shift ; do
      [ -z "$1" ] && break
      NAME=$1
      if test -e $CONFIG_DIR/$NAME.conf ; then
        log_daemon_msg "  Starting VPN '$NAME'"
        start_vpn
      else
        log_failure_msg "  Starting VPN '$NAME': missing $CONFIG_DIR/$NAME.conf file !"
       STATUS=1
      fi
    done
  fi
  exit ${STATUS:-0}
  ;;
stop)
  log_action_begin_msg "Stopping $DESC"
  if test -z "$2" ; then
    PIDFILE=
    for PIDFILE in `ls /var/run/openvpn.*.pid 2> /dev/null`; do
      NAME=`echo $PIDFILE | cut -c18-`
      NAME=${NAME%%.pid}
      log_daemon_msg "  Stopping VPN '$NAME'"
      stop_vpn
    done
    if test -z "$PIDFILE" ; then
      log_warning_msg "  No VPN is running."
    fi
  else
    while shift ; do
      [ -z "$1" ] && break
      if test -e /var/run/openvpn.$1.pid ; then
        log_daemon_msg "  Stopping VPN '$1'"
        PIDFILE=`ls /var/run/openvpn.$1.pid 2> /dev/null`
        NAME=`echo $PIDFILE | cut -c18-`
        NAME=${NAME%%.pid}
        stop_vpn
      else
        log_failure_msg "  Stopping VPN '$1': No such VPN is running."
      fi
    done
  fi
  ;;
# Only 'reload' running VPNs. New ones will only start with 'start' or 'restart'.
reload|force-reload)
  log_action_begin_msg "Reloading $DESC"
  PIDFILE=
  for PIDFILE in `ls /var/run/openvpn.*.pid 2> /dev/null`; do
    NAME=`echo $PIDFILE | cut -c18-`
    NAME=${NAME%%.pid}
# If openvpn if running under a different user than root we'll need to restart
    if egrep '^[[:blank:]]*user[[:blank:]]' $CONFIG_DIR/$NAME.conf > /dev/null 2>&1 ; then
      log_daemon_msg "  Stopping VPN '$NAME'"
      stop_vpn
      sleep 1
      log_daemon_msg "  Restarting VPN '$NAME'"
      start_vpn
    else
      log_daemon_msg "  Restarting VPN '$NAME'"
      kill -HUP `cat $PIDFILE` || true
      log_end_msg 0
    fi
  done
  if test -z "$PIDFILE" ; then
    log_warning_msg "  No VPN is running."
  fi
  ;;
# Only 'soft-restart' running VPNs. New ones will only start with 'start' or 'restart'.
soft-restart)
  log_action_begin_msg "Soft-restarting $DESC"
  PIDFILE=
  for PIDFILE in `ls /var/run/openvpn.*.pid 2> /dev/null`; do
    NAME=`echo $PIDFILE | cut -c18-`
    NAME=${NAME%%.pid}
    log_daemon_msg "  Soft-restarting VPN '$NAME'"
    kill -USR1 `cat $PIDFILE` || true
    log_end_msg 0
  done
  if test -z "$PIDFILE" ; then
    log_warning_msg "  No VPN is running."
  fi
  ;;
restart)
  shift
  $0 stop ${@}
  sleep 1
  $0 start ${@}
  ;;
cond-restart)
  log_action_begin_msg "Restarting $DESC"
  PIDFILE=
  for PIDFILE in `ls /var/run/openvpn.*.pid 2> /dev/null`; do
    NAME=`echo $PIDFILE | cut -c18-`
    NAME=${NAME%%.pid}
    log_daemon_msg "  Stopping VPN '$NAME'"
    stop_vpn
    sleep 1
    log_daemon_msg "  Restarting VPN '$NAME'"
    start_vpn
  done
  if test -z "$PIDFILE" ; then
    log_warning_msg "  No VPN is running."
  fi
  ;;
status)
  GLOBAL_STATUS=0
  if test -z "$2" ; then
    # We want status for all defined VPNs.
    # Returns success if all autostarted VPNs are defined and running
    if test "x$AUTOSTART" = "xnone" ; then
      # Consider it a failure if AUTOSTART=none
      log_warning_msg "No VPN autostarted"
      GLOBAL_STATUS=1
    else
      if ! test -z "$AUTOSTART" -o "x$AUTOSTART" = "xall" ; then
        # Consider it a failure if one of the autostarted VPN is not defined
        for VPN in $AUTOSTART ; do
          if ! test -f $CONFIG_DIR/$VPN.conf ; then
            log_warning_msg "VPN '$VPN' is in AUTOSTART but is not defined"
            GLOBAL_STATUS=1
          fi
        done
      fi
    fi
    for CONFIG in `cd $CONFIG_DIR; ls *.conf 2> /dev/null`; do
      NAME=${CONFIG%%.conf}
      # Is it an autostarted VPN ?
      if test -z "$AUTOSTART" -o "x$AUTOSTART" = "xall" ; then
        AUTOVPN=1
      else
        if test "x$AUTOSTART" = "xnone" ; then
          AUTOVPN=0
        else
          AUTOVPN=0
          for VPN in $AUTOSTART; do
            if test "x$VPN" = "x$NAME" ; then
              AUTOVPN=1
            fi
          done
        fi
      fi
      if test "x$AUTOVPN" = "x1" ; then
        # If it is autostarted, then it contributes to global status
        status_of_proc -p /var/run/openvpn.${NAME}.pid openvpn "VPN '${NAME}'" || GLOBAL_STATUS=1
      else
        status_of_proc -p /var/run/openvpn.${NAME}.pid openvpn "VPN '${NAME}' (non autostarted)" || true
      fi
    done
  else
    # We just want status for specified VPNs.
    # Returns success if all specified VPNs are defined and running
    while shift ; do
      [ -z "$1" ] && break
      NAME=$1
      if test -e $CONFIG_DIR/$NAME.conf ; then
        # Config exists
        status_of_proc -p /var/run/openvpn.${NAME}.pid openvpn "VPN '${NAME}'" || GLOBAL_STATUS=1
      else
        # Config does not exist
        log_warning_msg "VPN '$NAME': missing $CONFIG_DIR/$NAME.conf file !"
        GLOBAL_STATUS=1
      fi
    done
  fi
  exit $GLOBAL_STATUS
  ;;
*)
  echo "Usage: $0 {start|stop|reload|restart|force-reload|cond-restart|soft-restart|status}" >&2
  exit 1
  ;;
esac

exit 0

# vim:set ai sts=2 sw=2 tw=0:
